core/chains/evm/logpoller/orm.go

```diff
@@ -142,6 +146,27 @@ func (o *ORM) DeleteLogsAfter(start int64, qopts ...pg.QOpt) error {
 	return q.ExecQ(`DELETE FROM evm_logs WHERE block_number >= $1 AND evm_chain_id = $2`, start, utils.NewBig(o.chainID))
 }
 
+type Exp struct {
+	Address      common.Address
+	EventSig     common.Hash
+	Expiration   time.Time
+	TimeNow      time.Time
+	ShouldDelete bool
+}
+
+func (o *ORM) DeleteExpiredLogs(qopts ...pg.QOpt) error {
+	q := o.q.WithOpts(qopts...)
+
+	return q.ExecQ(`WITH r AS
+				( SELECT address, event, MAX(retention) AS retention
+					FROM evm_log_poller_filters WHERE evm_chain_id=$1 
+					GROUP BY evm_chain_id,address, event HAVING NOT 0 = ANY(ARRAY_AGG(retention))
+				) DELETE FROM evm_logs l USING r
+					WHERE l.evm_chain_id = $1 AND l.address=r.address AND l.event_sig=r.event
+						AND l.created_at + (r.retention / 10^9 * interval '1 second') <= STATEMENT_TIMESTAMP()`, // retention is in nanoseconds (time.Duration aka BIGINT)
```

Comment:
Maybe dedent it so its not so far over to the right?

Reply:
Has this been benchmarked/profiled?

Reply:
I haven't profiled it.   I guessed we'd need to do it on a node that's connected to a testnet, maybe run a long replay first to make sure it saves a large number of log events.

The WITH part of it should be negligible, because the `evm_log_pollers_filters` table will be much smaller than `evm_logs` (aside from initially, when they are both small).  The DELETE itself could take a while if there are a lot of old logs.  By default this will run once every 2400 blocks... so as long as it's running regularly on schedule, it shouldn't have more than a few thousand rows to delete each time.  I should at least check to make sure it's using an index.

Oh, that reminds me... I forgot to increase the default timeout for this query!  We should definitely do that.

Reply:
Oh, it doesn't use an index actually:
```
 Delete on evm_logs l  (cost=0.05..0.06 rows=0 width=0)
   ->  Hash Join  (cost=0.05..0.06 rows=1 width=143)
         Hash Cond: ((l.evm_chain_id = r.evm_chain_id) AND (l.address = r.address) AND (l.event_sig = r.event))
         Join Filter: ((l.created_at + (((r.retention)::double precision / '1000000000'::double precision) * '00:00:01'::interval)) <= now())
         ->  Seq Scan on evm_logs l  (cost=0.00..0.00 rows=1 width=119)
         ->  Hash  (cost=0.03..0.03 rows=1 width=250)
               ->  Subquery Scan on r  (cost=0.01..0.03 rows=1 width=250)
                     ->  HashAggregate  (cost=0.01..0.02 rows=1 width=113)
                           Group Key: evm_log_poller_filters.evm_chain_id, evm_log_poller_filters.address, evm_log_poller_filters.event
                           ->  Seq Scan on evm_log_poller_filters  (cost=0.00..0.00 rows=1 width=113)
```
The `Hash Joins`'s will all be fast, but I think `Join Filter` on `created_at` could be improved.  That column doesn't have an index defined on it, but I could add one.  Even without an index, I think the table would have to get pretty large before it mattered... but given what happened with the pipeline_runs table seems worth being proactive about.